<a href="https://colab.research.google.com/github/xpertdesh/ml-class21/blob/main/projects/bach_chorales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center>Bach Chorales Project</center>
---
<br>
This project consists of predicting the chords that Bach used based off from a set of notes. Johann Sebastian Bach was a German composer and musician of the Baroque period.<br> 
Fun fact: This is the second time in my life hearing about Bach. First time, was for 326 when Stephen made us read that humungous book GEB.<br><br>
Before looking at the data set. I understand there are 3 columns of actual data, and 2 that wouldn't be used in the project.

### Ok let's get started!


https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip

*   First,  we will download and unzip the file
*   Then, we will import the neccessary libraries in order to take a look at the data
*   Finally, we will explore and analyze the data


In [2]:
! wget https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip

--2021-02-25 15:56:29--  https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 41761 (41K) [application/zip]
Saving to: ‘bach.zip’

bach.zip            100%[===================>]  40.78K  --.-KB/s    in 0.004s  

2021-02-25 15:56:29 (9.66 MB/s) - ‘bach.zip’ saved [41761/41761]



In [3]:
! unzip bach.zip -d bach_chorales

Archive:  bach.zip
  inflating: bach_chorales/bach.csv  


The previous step actually took me quite a while but that's fine.<br>
Ok, with that out of the way. We can now import the neccessary libraries.<br><br><br>

##1. Import Neccessary Libraries
<br><br><br>

In [4]:
#data analysis libraries 
import numpy as np
import pandas as pd

#visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

#ignore warnings
import warnings
warnings.filterwarnings('ignore')

<br><br><br>

##2. Read in and Explore the Data
It's time to read in the data using `pd.readcsv()`, and take a first look at the data using the `describe()` function.<br><br><br>

In [5]:
bach = pd.read_csv('bach_chorales/bach.csv')
bach.describe(include='all')

choral_ID  event_number     C    C#  ...     B  bass        meter chord_label
count       5665   5665.000000  5665  5665  ...  5665  5665  5665.000000        5665
unique        60           NaN     2     2  ...     2    16          NaN         102
top     002908ch           NaN    NO    NO  ...    NO     D          NaN         D_M
freq         207           NaN  3875  4711  ...  3874   689          NaN         503
mean         NaN     53.374404   NaN   NaN  ...   NaN   NaN     3.134863         NaN
std          NaN     37.268208   NaN   NaN  ...   NaN   NaN     1.109710         NaN
min          NaN      1.000000   NaN   NaN  ...   NaN   NaN     1.000000         NaN
25%          NaN     24.000000   NaN   NaN  ...   NaN   NaN     2.000000         NaN
50%          NaN     48.000000   NaN   NaN  ...   NaN   NaN     3.000000         NaN
75%          NaN     75.000000   NaN   NaN  ...   NaN   NaN     4.000000         NaN
max          NaN    207.000000   NaN   NaN  ...   NaN   NaN     5.000000         NaN

[11 rows x 17 columns]

<br><br><br>

#3. Data Analysis
We're going to consider the features in the dataset and how complete they are.
<br><br><br>

In [6]:
#get of a list of the features within the dataset
bach.columns

Index(['choral_ID', 'event_number', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G',
       'G#', 'A', 'A#', 'B', 'bass', 'meter', 'chord_label'],
      dtype='object')

In [8]:
#see a sample of the data set to get an idea of the variables
bach.sample(10)

choral_ID  event_number    C  C#    D  ...   A#    B bass meter chord_label
1450  001907ch             9  YES  NO   NO  ...   NO   NO    C     4         A_m
1062  001306b_            17  YES  NO   NO  ...   NO   NO    F     4         F_M
960   001207b_             6   NO  NO   NO  ...  YES   NO    G     3         EbM
199   000206b_            38   NO  NO  YES  ...  YES   NO    G     5         G_m
928   001106b_           121   NO  NO  YES  ...   NO  YES    B     3         B_m
1066  001306b_            21  YES  NO   NO  ...   NO   NO    F     4         F_M
995   001207b_            41   NO  NO  YES  ...  YES   NO   Bb     2         BbM
3600  012606b_            25   NO  NO  YES  ...   NO   NO    F     2         D_m
2639  003907b_            16   NO  NO   NO  ...  YES   NO    G     3         EbM
2286  003206b_            42   NO  NO  YES  ...   NO  YES    B     5         G_M

[10 rows x 17 columns]

<br><br>


* **Numerical Features:** Event Number(Discrete), Meter(Discrete)
* **Categorical Features:** Western Scale(Discrete), Bass(Discrete)
* **Label:** Chord Label(Discrete)

<br>Choral Id, and Event Number will not be used for training.
<br><br>
Let's take another look at the dataset, now that we know what kinds of features we're working with.<br><br><br>


In [ ]:
#see a summary of the dataset
bach.describe(include='all')

choral_ID  event_number     C    C#  ...     B  bass        meter chord_label
count       5665   5665.000000  5665  5665  ...  5665  5665  5665.000000        5665
unique        60           NaN     2     2  ...     2    16          NaN         102
top     002908ch           NaN    NO    NO  ...    NO     D          NaN         D_M
freq         207           NaN  3875  4711  ...  3874   689          NaN         503
mean         NaN     53.374404   NaN   NaN  ...   NaN   NaN     3.134863         NaN
std          NaN     37.268208   NaN   NaN  ...   NaN   NaN     1.109710         NaN
min          NaN      1.000000   NaN   NaN  ...   NaN   NaN     1.000000         NaN
25%          NaN     24.000000   NaN   NaN  ...   NaN   NaN     2.000000         NaN
50%          NaN     48.000000   NaN   NaN  ...   NaN   NaN     3.000000         NaN
75%          NaN     75.000000   NaN   NaN  ...   NaN   NaN     4.000000         NaN
max          NaN    207.000000   NaN   NaN  ...   NaN   NaN     5.000000         NaN

[11 rows x 17 columns]

<br><br><br>

**Some Observatios:** 

* There are a total of 5,665 chords in our dataset.
* All the data is complete, no missing values in the dataset.
* We're going to have to hot-encode the Western Scale, and Bass columns.
* Choral Id and Event Number columns will get dropped since they won't help the fitting process.
<br><br>

**Some Predictions:** 
* I don't know anything about chords to be specific, boring lol.
* Maybe I'll ask a friend to predictions from someone who knows.
<br><br><br>

# 4. Data Visualization

Time to visualize some data.

In [43]:
#draw a barplot of meter values
#sns.barplot(x="chord_label", y="meter", data=bach)

#print percentages of meters
#print("Meter:\n", bach["meter"].value_counts(normalize = True)*100)
print("Bass:\n", bach["meter"][bach["meter"] == 2].value_counts(normalize = True)*100)

Bass:
 2    100.0
Name: meter, dtype: float64


According to the graph, it would seem that the "5" meter is not used a lot when making chords.
<br><br><br>

In [38]:
#draw a barplot of bass values
#sns.barplot(x="chord_label", y="bass", data=bach)

#print percentages of bass
print("Bass:\n", bach["bass"].value_counts(normalize = True)*100)

Bass:
 D     12.162401
A     12.144748
G     12.021183
E     11.968226
C     10.591350
F      8.155340
F#     8.049426
B      7.925861
Bb     5.101500
C#     3.318623
Eb     3.195057
G#     2.594881
D#     1.235658
Ab     0.600177
Db     0.529568
A#     0.406002
Name: bass, dtype: float64


Although, we can't plot the remaining of the columns because they're not numeric, we can still see the percentages for each columns. The most commonly used basses are "D, A, G, E, C, F", which is roughly 60% of the bass column.
<br><br><br>
So, we're not going to plot the 12 notes of a Western Scale because the percentage doesn't show if those notes were actually useful to creating a chord. The percentages are not related to the chord_labels.
<br><br><br>

#5. Cleaning Data

Time to clean our data, hot-encode the bass and notes columns, and delete the unnecessary columns.
<br><br><br>